File: cronenberg/importer/graphite.py
import requests
import logging
import json
import inflection
import urllib
from cronenberg.model import *
from cronenberg import app, db

log = logging.getLogger(__name__)
mgr = database.DatabaseManager(db)

class GraphiteDashboardImporter(object):
    def __init__(self, url):
        self.url = url

    def get_dashboard_names(self, query=''):
        response = requests.post('{0}/dashboard/find/'.format(self.url),
                                 params={'query':query})
        return [ d['name'] for d in response.json()['dashboards'] ]

    def get_dashboard(self, name):
        response = requests.get('{0}/dashboard/load/{1}'.format(self.url, name))
        return response.json()['state']

    def dump_dashboards(self, query):
        names = self.get_dashboard_names(query)
        print json.dumps([ self.get_dashboard(n) for n in names ], cls=EntityEncoder, indent=4)

    def import_dashboards(self, query, **kwargs):
        names = self.get_dashboard_names(query)
        log.info('Found {0} dashboards to import'.format(len(names)))
        log.info(','.join(names))

        for name in names:
            log.info('Importing {0}'.format(name))
            dash = self.import_dashboard(name, **kwargs)
            mgr.store_dashboard(dash)

    def import_dashboard(self, name, **kwargs):
        return self.__convert(self.get_dashboard(name), **kwargs)

    def __convert(self, graphite_dashboard, fluid=False, columns=2):
        span = 12 / columns
        name = graphite_dashboard['name']
        dashboard = database.Dashboard(title=inflection.parameterize(name),
                                       category='Graphite',
                                       tags=[database.Tag('imported'), database.Tag('from:graphite-web')],
                                       imported_from = '{0}/dashboard/{1}'.format(app.config['GRAPHITE_URL'], urllib.quote(name)))
        definition = DashboardDefinition()
        section = Section(is_container=not fluid)
        definition.items.append(section)

        # if 'defaultGraphParams' in graphite_dashboard:
        #     default_width = graphite_dashboard['defaultGraphParams'].get('width', None)
        #     print graphite_dashboard['defaultGraphParams']
        #     if default_width and default_width > 800:
        #         num_columns = 1
        #         span = 12

        row = Row()
        for graph in graphite_dashboard['graphs']:
            # Graphite's dashboard API is so redundant. Each graph is
            # a 3-element array:
            # [
            #    targets array,
            #    options dict (which contains the targets array too),
            #    render URL string
            #  ]
            targets, options, render_url = graph
            presentation = None
            stacked_p = render_url.find('stacked') != -1 or options.get('areaMode', None) == 'stacked'
            query_name = 'q' + str(len(definition.queries))
            targets = options.get('target', [])
            definition.queries[query_name] = targets[0] if len(targets) == 1 else targets
            if stacked_p:
                presentation = StackedAreaChart(query_name=query_name, title=options.get('title', ''))
            else:
                presentation = StandardTimeSeries(query_name=query_name, title=options.get('title', ''))
            presentation.options['yAxisFormat'] = ',.2s'
            presentation.height = 4
            if 'template' in options:
                presentation.options['palette'] = options['template']
            if 'vtitle' in options:
                presentation.options['yAxisLabel'] = options['vtitle']
                presentation.options['yShowMaxMin'] = True
            presentation.options['margin'] = { 'top' : 16, 'left' : 80, 'right' : 0, 'bottom' : 16}

            row.items.append(Cell(span=span, items=presentation))
            if len(row.items) == columns:
                section.items.append(row)
                row = Row()

        dashboard.definition = database.DashboardDef(definition=dumps(definition))
        return dashboard

File: tessera-frontend/src/ts/importer/graphite.ts
import { manager, config } from '../app'
import * as models from '../models'
import { logger } from '../util'
import * as charts from '../charts'

declare var require
const axios      = require('axios')
const URI        = require('urijs')
const inflection = require('inflection')

const log = logger('importer.graphite')

function translate(gdash: any, opt : any = {}) : models.Dashboard {
  return manager.without_updates(() => {
    let columns = opt.columns || 3
    let span = 12 / columns

    let section = new models.Section()
    let definition = new models.DashboardDefinition({
      items: [section]
    })

    let dash = new models.Dashboard({
      title: inflection.dasherize(gdash.name),
      category: 'Graphite',
      tags: ['imported'],
      definition: definition
    })

    let row = new models.Row()
    for (let graph of gdash.graphs) {
      let [targets, options, render_url] = graph
      let is_stacked = render_url.indexOf('stacked') > -1 || (options.areaMode && options.areaMode === 'stacked')
      let query_name = 'q' + Object.keys(definition.queries).length
      let query = new models.Query({
        name: query_name,
        targets: options.target || []
      })
      definition.add_query(query)

      let presentation = new models.StandardTimeSeries({
        query: query,
        title: options.title,
        stack_mode: is_stacked ? charts.StackMode.NORMAL : charts.StackMode.NONE,
        options: {
          y1: {
            label: options.vtitle,
            max: options.yMax,
            min: options.yMin
          },
          palette: options.template
        }
      })

      row.add(new models.Cell({
        span: span,
        items: [presentation]
      }))

      if (row.items.length === columns) {
        section.add(row)
        row = new models.Row()
      }
    }

    if (row && row.items && row.items.length) {
      section.items.push(row)
    }

    return dash
  })
}

/**
 * Translate from a Graphite dashboard UI URL to an API URL to load
 * the JSON definition.
 */
export function get_api_url(dash_url: string) : string {
  let uri = new URI(dash_url)
  return uri.segment('load')
    .segment(uri.fragment())
    .fragment(null)
    .toString()
}

/**
 * Fetch a Graphite dashboard definition, convert it to a Tessera
 * dashboard definition, and upload it to the Tessera server via the
 * API.
 */
export async function import_dashboard(url: string) : Promise<any> {
  let api_url = get_api_url(url)
  log.info(`Importing dashboard from ${api_url}...`)

  // TODO: this is duplicated from query.ts
  let axios_config : any = {
    url: api_url.toString(),
    method: 'get'
  }

  let auth = config.GRAPHITE_AUTH ? config.GRAPHITE_AUTH.trim() : ''
  if (auth) {
    let [username, password] = auth.split(':')
    axios_config.auth = {
      username: username,
      password: password
    }
    axios_config.withCredentials = true
  }

  return axios(axios_config)
    .then(response => translate(response.data.state))
    .then(dash => {
      return manager.client.dashboard_create(dash)
        .then(response => {
          return response
        })
        .catch(error => {
          manager.error(`Error creating dashboard. ${error}`)
        })
    })
}
